README.md
# An example of a BARS standalone app

This repository hosts an example of a BARS app - a compiled program that we advise should be used instead of ROOT macros.

## Requirements

This app only requires a BARS installation. 
The exact steps are described [here](https://bars-docs.jinr.ru/install).

## Installation and compilation 

1. Clone this repository to a directory named as this app

```
git clone git@git.jinr.ru/Baikal/bars-app-standalone-example.git my-bars-app
```

2. Create a build directory.

```
cd my-bars-app
mkdir build
```

3. Create a Makefile and compile

```
cd build
cmake ..
make
```

4. Install it

```
make install
```

This will put an output binary into `~/.local/bin/bars`. Make sure `~/.local/bin` is  in your `$PATH`!.


4. Start your app!

```
my-bars-app
```


releaseNotes.txt
12.02.2019 - git commit ---
	* Changes in the folder paths for VM68 (processExperimental, json, saveHistograms)
	* 2019 geometry read directly from /home/fajtak/geometry-tmp/

12.02.2019 - git commit aa3740c13ad24331de52754d0e1626afb29f3366
	* Huge change!!! All inputs (data, MC, MCcascades) are now processed in the same way
	* standardized UniformEvent was implemented and all different inputs are transformed to this format
	* DoTheMagicUnified()
	* EstimatePosition based on the Matrix approach from ANTARES
	* Caussality filter
	* In the next commit all functions for individual inputs are going to be removed
	* QMEstimatoCut decreased from 2000 to 600


11.02.2020 - git commit 108c4fbb0d926d85774b8827724a57e1d2fc2842
	* QCutOverall added with default value 500
	* QCutHits increased to 10
	* NCut decreased to 70
	* logLikelihood return value divided by g_pulses size not 288
	* Now the first non zero time of the run is found

04.02.2020 - git commit 0414a2e6ba3d8b026c26beb7928e85ab014656d8
	* minor changes to studyReconstructed(MC)Cascades.C
	* config.rc: number of hits increased from 60 -> 80 + TDelayCut 400 -> 600
	* IMPORTANT: MCCascades and all corresponding filters extended for MC noise
	* h_chargePerEvent histogram added and this property is also used to filter events (overallEventCharge [p.e] > 250)
	* error in the MCCascade reconstruction corrected
	* event visualization for MCCascade added
	* PrintCascadeJSON() added to save jsons that can be visualized by the event viewer
	* structure of the BARS::DATA::Direcotry and FileName has changed
	* the "o" switch now can be used to save results to other folder (useful for VM processing)
	* Estimate initial position again based on the grid approach

28.01.2020 - git commit 13123210d57a882fac10f7d1bbaa1a935139ba62
	* The nonHit term added to the logLikelihood fitting
	* the initial position and time estimation again based on the highest Q OM
	* charge visualization extended to show expected charge on all OMs
	* initDist and FinalDist added to the nTuples for MC data
	* time of processing added

07.01.2020 - git commit 41a7cc590aca373748dffa028379057e85b296fe
	* The first fit is again done with MEstimator

07.01.2020 - git commit 59abf51237727bee7d64a0d359a150a18885dab5
	* SetFCN(chi2) added before the first fit

06.01.2020 - git commit 5de116f1f39ea6afe1c0239300e9856d28edb4f4
	* new switch "a" to process Zhan-Arys MC cascades
	* values in the config.rc changed and branchCut and likelihoodCut values added
	* in studyReconstructedMCCascades.C script added a new histograms for position mismatch
	* added config file configDZH.rc for processing of the simulated cascades
	* logLikelihood fitting CORRECTED!!!
	* the first fit in the chain is again chi2
	* to guess initial fitiing time and position a grid of positions around detector are tested and value with smallest chi2 is used.
	* added all the filters for MCcascades
	* New file with MCcascade geometry added array2016_phys_kebkal.dat

11.12.2019 - git commit beab71ada52335628d4c038c20800208197f2902
	* new switch "u" to process MC single up-going atmospheric neutrinos
	* added two new bash scripts to download and read MC single up-going atmospheric neutrinos (downloadDataNu and readMCNu)
	* added function OMIDAlreadyInGPulses for MC data

10.12.2019 - git commit 45278b764bf40181994348579a0ee4cec64fa917
	* Huge extension of the whole program. Now MC atmospheric muon bundles can be read and cascades created on the muon tracks can be studied.
	* Added switch -m with no argument which enables to process MC data
	* Now the charge of all the pulses is calibrated -> QCut has to be set in p.e.
	* Added two bash scripts downloadMCData and readMC that downloads MC data and transforms them from wout to root structure
	* Added root script studyReconstructedMCCascades.C to study reconstructed MC cascades where all the parameters are known
	* g_pulses now have MCflag member that tels you the origin of the hit
	* scattering correction and Log(Q) removed from ExpectedTime -> It will be studied in the future
	* all the filters has been replicated to be able to work with BEvents from MC simulations
	* output of the program has been extended for RunInfo. Now the number of entries and time duration of the run is printed


04.12.2019 - git commit 3a305382d6f026cf422822e9ddd05ce7a577974a
	* Minor corrections

03.12.2019 - git commit 9f1a0d6847b07ca5d7a3e7136dba78e313a14401

	* default value of the LikelihoodCut increased from 2 -> 2.5
	* studyReconstructedCascades.C extended significantly. Now it shows also reconstructed energy and theta and also energy and TFilterHits cuts are used
	* TMultiGraphs in charge and event visualizations have axis labels now
	* existence of the 4D logLikelihood table is checked

29.11.2019 - git commit 3249e8e195299ca1276254f504876cef53d3f0fb
	
	* Loglikelihood for direction and energy reconstruction added
	* 4D pdf table is read
	* 8 string graphs with measured and expected charges are created

22.11.2019 - git commit 47fc8ce4b17b7e12e5ab351a816b852ac311ec7c

	* New switch "-t productionID" (tag) has been added to be able to change the name of the data processing production ID. The default value is "barsv051" other common option is "r01_i01_j01_t01"

20.11.2019 - git commit d0292fc96f7f60b4cb176654feaee6db1f1547dc

	* In downloadJoint and downloadJointAllRuns added also download of the qcalib file from Zhenya's DQM
	* Qcalib is read from h0XXX.qcalib file and if qcalib constant is -1 it is assumed that the OM is not working
	* CloseHitsFilter is now based on the qcalib and not on the h_nHitsOM
	* start time of the run is newly taken from the fifth header entry since there are runs where the first two entries have zero values

19.11.2019 - git commit db5a086f8e6ea6feec34a087b8a9e54c1963ae50

	* src/config.rc is now used as a default option for processAllRuns script
	* changes in the values of the cut variables saved in src/config.rc

src/opts.cc
#include "opts.h"

#include <iostream>
#include <getopt.h>
#include <cstdlib>
#include <stdlib.h>

#include <TEnv.h>

#include "BARS.h"

int gNEventsProcessed = -1; 
int gNCut = -1;
double gQCutOverall = -1;
double gQCut = -1;
int gQCutHits = -1;
double gQCutChi2 = -1;
double gTCutTimeWindowNs = -1;
int gTCutHits = -1;
double gTCutChi2 = -1;
double gZCut = -1;
double gTDelayCut = -1;
double gQRatioCut = -1;
int gBranchCut = -1;
int gVisEventID = -1;
double gScatteringCorrection = 10; //10 ns per 15 meters 
double gLikelihoodCut = -1;
bool gMCMu = false;
bool gMCNu = false;
bool gMCCas = false;
std::string gProductionID = "";

using namespace BARS;

// At the end of list you should add 'App::opt_NULL' element
static const struct App::ProgramOption_t options_list[]{
  // --help and --config options will be added automaticly
	{App::opt_Verbose, NOT_REQUIRED},
	{App::opt_Input,   NOT_REQUIRED},
	{App::opt_Output,  NOT_REQUIRED},
	{App::opt_Cluster, NOT_REQUIRED},
	{App::opt_Season,  NOT_REQUIRED},
	{App::opt_Run,     NOT_REQUIRED},
	{
		{
			"number", 'n',
			required_argument,
			"set number of events you want to process",
			[](char* argv) {gNEventsProcessed = atoi(argv);},
			[]() {;}
		},
		NOT_REQUIRED
	},
	{
		{
			"view", 'w',
			required_argument,
			"set ID of the event that should be visualized",
			[](char* argv) {gVisEventID = atoi(argv);},
			[]() {;}
		},
		NOT_REQUIRED
	},
	{
		{
			"tag", 't',
			required_argument,
			"set production ID of the data that will be processed",
			[](char* argv) {gProductionID = argv;},
			[]() {;}
		},
		NOT_REQUIRED
	},
	{
		{
			"mcMu", 'm',
			no_argument,
			"use mc atmospheric muon bundles data as the input for the program",
			[](char* argv) {gMCMu = true;},
			[]() {;}
		},
		NOT_REQUIRED
	},
	{
		{
			"mcNu", 'u',
			no_argument,
			"use mc up-going single muons data as the input for the program",
			[](char* argv) {gMCNu = true;},
			[]() {;}
		},
		NOT_REQUIRED
	},
	{
		{
			"mcCas", 'a',
			no_argument,
			"use mc cascade data from Zhan-Arys as the input for the program",
			[](char* argv) {gMCCas = true;},
			[]() {;}
		},
		NOT_REQUIRED
	},
  // !!! must be at the end
	{App::opt_NULL, NOT_REQUIRED}
};

// Read resource file, fill parameters
//_____________________________________________________________________
void readRC(const char* rcpath)
{
	TEnv env;    
	if (-1 == env.ReadFile(App::RCPath, kEnvAll)) {
		std::cerr << "ERROR: failed to read configuration file at '" << App::RCPath << "'" << std::endl;
		exit(1);
	}
	gNCut = env.GetValue("NCut", 70);
	gQCutOverall = env.GetValue("QCutOverall", 400.0);
	gQCut = env.GetValue("QCut", 15.0);
	gQCutHits = env.GetValue("QCutHits", 6);
	gQCutChi2 = env.GetValue("QCutChi2", 50);
	gTCutTimeWindowNs = env.GetValue("TCutTimeWindowNs", 50);
	gTCutHits = env.GetValue("TCutHits", 20);
	gTCutChi2 = env.GetValue("TCutChi2", 50);
	gZCut = env.GetValue("ZCut",600);
	gTDelayCut = env.GetValue("TDelayCut",400);
	gQRatioCut = env.GetValue("QRatioCut",80);
	gBranchCut = env.GetValue("BranchCut",0);
	gLikelihoodCut = env.GetValue("LikelihoodCut",2.5);
}

// Parse options passed to the application.
//_____________________________________________________________________
void parseOpts(int argc, char **argv)
{
	App::ParseProgramOptions(argc, argv, options_list);
}

//_____________________________________________________________________
void checkParams()
{
	App::CheckProgramOptions(options_list);
}